caddyhttp/httpserver/server.go

```diff
@@ -356,6 +356,12 @@ func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	c := context.WithValue(r.Context(), OriginalURLCtxKey, urlCopy)
 	r = r.WithContext(c)
 
+	// Setup a replacer for the request that keeps track of placeholder
+	// values across plugins.
+	replacer := NewReplacer(r, nil, "")
```

Comment:
Wanted to double-check: does the `log` middleware still set its own "empty" value (should default to `-` at least for the default log format)?

Reply:
Yep! That's correct. Every call to `NewReplacer` it a partial clone of the shared one that sets it's own `emptyValue`: https://github.com/kujenga/caddy/blob/per-request-replacer/caddyhttp/httpserver/replacer.go#L108

The log package still does that for each request: https://github.com/kujenga/caddy/blob/per-request-replacer/caddyhttp/log/log.go#L48

Reply:
Okay, cool. So isn't the next step to get the middlewares to pull the Replacer off the context instead of making their own? (I guess your code helps by using the context's one, but I imagine it'd be "cleaner" if they just pulled from the context directly?)